### ISSUE TITLE
Training with custom dataset

### ISSUE THREAD

**Issue by seongin0417** [aau-cns/poet#32]
Hi, I'm trying to train PoET with my own dataset.
And I've already trained mask-rcnn in same version with docker.

When I ran inference using the Mask R-CNN model I trained, it performed well(high scores, almost 1). but in your code’s MaskRCNNBackbone forward method the scores are coming out far too low(under 0.4).

How did you train your backbone? Could I get the code you trained it?

Here is my inference code to find out the issue.

import torch
from torch import nn
from torchvision.ops import MultiScaleRoIAlign
from torchvision.models.detection.anchor_utils import AnchorGenerator
from torchvision.models.detection.backbone_utils import resnet_fpn_backbone
from torchvision.models.detection.mask_rcnn import MaskRCNN
from torchvision.transforms import functional as F
from torchvision.transforms import ToTensor
from PIL import Image
from collections import OrderedDict
from typing import Dict, List, Optional
from torchvision.models.detection.rpn import AnchorGenerator, concat_box_prediction_layers

from torchvision.models.detection.image_list import ImageList
from torchvision.models.detection._utils import overwrite_eps
# from util.misc import (NestedTensor, nested_tensor_from_tensor_list)

class MaskRCNNBackbone(MaskRCNN):
    def __init__(
        self,
        n_classes: int = 2,
        return_interm_layers=True,
        train_backbone=False
    ):
        backbone = resnet_fpn_backbone('resnet50', pretrained=False)
        self.train_backbone = train_backbone

            self.return_layers = ['2', '3', 'pool']
            self.strides = [8, 16, 32]
            self.num_channels = [256, 256, 256]
        else:
            self.return_layers = ['pool']
            self.strides = [256]
            self.num_channels = [256]
        
        super().__init__(backbone=backbone,
                         num_classes=n_classes)

    def forward(self, tensor_list: ImageList):
        image_sizes = [img.shape[-2:] for img in tensor_list.tensors]
        features = self.backbone(tensor_list.tensors)

        feature_maps = list(features.values())
        objectness, pred_bbox_deltas = self.rpn.head(feature_maps)
        grid_sizes = [f.shape[-2:] for f in feature_maps]
        image_size = tensor_list.tensors.shape[-2:]
        dtype, device = feature_maps[0].dtype, feature_maps[0].device

        strides = [
            [torch.tensor(image_size[0]//g[0], dtype=torch.int64, device=device),
             torch.tensor(image_size[1]//g[1], dtype=torch.int64, device=device)]
            for g in grid_sizes
        ]
        self.rpn.anchor_generator.set_cell_anchors(dtype, device)
        all_anchors = self.rpn.anchor_generator.grid_anchors(grid_sizes, strides)
        anchors = [
            torch.cat(all_anchors) for _ in range(len(tensor_list.tensors))
        ]

        objectness, pred_bbox_deltas = concat_box_prediction_layers(objectness, pred_bbox_deltas)
        proposals = self.rpn.box_coder.decode(pred_bbox_deltas.detach(), anchors)
        proposals = proposals.view(len(anchors), -1, 4)
        boxes, scores = self.rpn.filter_proposals(proposals, objectness, image_sizes, [a.shape[0] for a in all_anchors])

        detections, _ = self.roi_heads(features, boxes, image_sizes)

        predictions = []
        for det in detections:
            preds = []
            for c, lbl in enumerate(det["labels"]):
                b = det["boxes"][c]
                score = det["scores"][c]
                preds.append(torch.hstack((b, score.unsqueeze(0), lbl.unsqueeze(0))))
            predictions.append(torch.stack(preds) if preds else None)

        out: Dict[str, ImageList] = {}
        for lvl in self.return_layers:
            feat = features[lvl]
            mask = tensor_list.mask
            resized_mask = F.interpolate(mask[None].float(), size=feat.shape[-2:])[0].to(torch.bool)
            out[lvl] = ImageList(feat, resized_mask)

        return predictions, out

if __name__ == "__main__":
    CHECKPOINT_PATH = "/output/maskrcnn/maskrcnn_epoch_015.pth"

    DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model = MaskRCNNBackbone(n_classes=2)
    model.load_state_dict(torch.load(CHECKPOINT_PATH, map_location=DEVICE))
    model.to(DEVICE).eval()

    img_path = "/data/test_all/rgb_sample/009000.png"
    image = Image.open(img_path).convert("RGB")
    img_tensor = ToTensor()(image).to(DEVICE)

    image_list, _ = model.transform([img_tensor])
    image_list = image_list.to(DEVICE)

    with torch.no_grad():
        detections, features = model(image_list)

    print("Detections per image:")
    for det in detections:
        print(det)

    print("\nIntermediate feature maps:")
    for lvl, img_list in features.items():
        print(f"Level {lvl}: tensor shape = {img_list.tensors.shape}")

**Comment by seongin0417** [comment]
No, the performance drop did not occur in my attached code — it happened in your code.

I proceeded as follows:
1. Train a Mask R-CNN model using my attached code on a custom dataset.
2. Perform inference using the weights from step 1 — achieved good performance.
3. Train and inference PoET with my backbone weights, but it struggled to produce accurate 2D bounding boxes.

**Comment by tgjantos** [comment]
Alright, I see. Can you provide me the parameter configuration that you used for PoET?

Best,
Thomas

**Comment by seongin0417** [comment]
Here is configuration

  --n_classes 1\
  --inference \
  --backbone maskrcnn \
  --backbone_cfg /opt/project/configs/workpiece31_rcnn.yaml \
  --bbox_mode backbone \
  --batch_size 8 \
  --num_queries 1 \
  --backbone_weights /opt/project/weights/maskrcnn_epoch_015.pth


and workpiece31_rcnn.yaml

anchor_sizes:
-
  - 32
-
  - 64
-
  - 128
-
  - 256
-
  - 512
backbone_str: resnet50-fpn
input_resize:
- 1080
- 1920
label_to_category_id:
  background: 0
  obj_000001: 1


All other settings were left as default.
The input image I used was an RGB image with a resolution of 1080×1920.
I trained Mask R-CNN using only a single object class labeled as class 1.


I really appreciate your support.

**Comment by seongin0417** [comment]
I have a question: when I look at the forward function of MaskRCNNBackbone, it appears that the raw image—not the preprocessed image (via GeneralizedRCNNTransform())—is being passed in. Could you please check this? Also, I’d appreciate it if you could let me know whether adding preprocessing would affect PoET output.

**Comment by tgjantos** [comment]
In general PoET performs also data preprocessing, see [here](https://github.com/aau-cns/poet/blob/main/data_utils/pose_dataset.py). It might be that your data preprocessing differs from the one applied by PoET. This would affect the performance of PoET. However, in inference mode no data processing is applied and the raw image as a tensor is given to the backbone. I am not familiar with GeneralizedRCNNTransform(), I also did not train with this specific pre-processing. 

Unfortunately, I am out of office for the next two weeks with no access to any computer capable of running any code. I will gladly give guidance. After my trip, I can also debug code. 

Anyway, I would suggest to compare your code to PoET's code for when the image is passed to the backbone. 

Best,
Thomas
